cranelift-frontend/src/ssa.rs

```diff
@@ -527,31 +565,39 @@ impl SSABuilder {
     fn finish_predecessors_lookup(
         &mut self,
         func: &mut Function,
-        temp_arg_val: Value,
-        temp_arg_var: Variable,
+        sentinel: Value,
+        var: Variable,
         dest_ebb: Ebb,
     ) {
         let mut pred_values: ZeroOneOrMore<Value> = ZeroOneOrMore::Zero;
 
-        // Iterate over the predecessors.
-        for _ in 0..self.predecessors(dest_ebb).len() {
-            // For each predecessor, we query what is the local SSA value corresponding
-            // to var and we put it as an argument of the branch instruction.
-            let pred_val = self.results.pop().unwrap();
+        let num_results = self.results.len();
+        let num_predecessors = self.predecessors(dest_ebb).len();
+
+        // Determine how many predecessors are yielding unique, non-temporary Values.
+        for i in 0..num_predecessors {
+            let pred_val = self.results[num_results - 1 - i];
             match pred_values {
                 ZeroOneOrMore::Zero => {
-                    if pred_val != temp_arg_val {
+                    if pred_val != sentinel {
                         pred_values = ZeroOneOrMore::One(pred_val);
                     }
                 }
                 ZeroOneOrMore::One(old_val) => {
-                    if pred_val != temp_arg_val && pred_val != old_val {
+                    if pred_val != sentinel && pred_val != old_val {
                         pred_values = ZeroOneOrMore::More;
+                        break;
                     }
                 }
-                ZeroOneOrMore::More => {}
+                ZeroOneOrMore::More => {
+                    break;
```

Comment:
If we're trying to keep the logic the same as before, don't these `break`s change it?

Reply:
The logic should be the same, unless I have misread something -- the multiple `pop()` calls were moved to a single `truncate()` below.

Reply:
Yeah, what I mean is that before the loop would keep iterating and `pred_values` could have been set more than once--which now that I look at it seems like a bug since it is a `sentinel` that guards this. Perhaps this fixes an unlikely bug?